### PR TITLE
frontend: add reporter facility

### PR DESCRIPTION
**Pull request by elena-pudovkina**
Create a reporter facility for reporting errors and warnings in .wake files.
Change parser.cpp to use the facility for errors with one location.
Change main.cpp to define the reporter object.

### CODE REVIEW

**Comment by terpstra** [comment]
Please merge #671 and then rebase this on top of master. It's hard to review when combined.

File: src/cli/main.cpp [hunk 1]
```diff
@@ -52,6 +53,14 @@
 #define TOSTRING(x) STRINGIFY(x)
 #define VERSION_STR TOSTRING(VERSION)
 
+class TerminalReporter : public DiagnosticReporter {
+  public:
+    void report(Diagnostic diagnostic) {
+      std::cerr << diagnostic.getMessage() << std::endl;
+    }
+};
+DiagnosticReporter *reporter = new TerminalReporter();
```
Comment (terpstra) [inline src/cli/main.cpp:62]: Does this need to be a pointer? I'd much rather just see:

```
TerminalReporter reporter;
```

Or does that not work?

File: src/frontend/diagnostic.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    const Severity severity;
+    const std::string message;   
+};
+
+class DiagnosticReporter {
+  public:
+    virtual void report(Diagnostic diagnostic) = 0;
+};
+
+#endif
```
Comment (terpstra) [inline src/frontend/diagnostic.h:29]: Missing newline on this file

File: src/frontend/diagnostic.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  private:
+    const Location location;
+    const Severity severity;
+    const std::string message;   
+};
+
+class DiagnosticReporter {
+  public:
+    virtual void report(Diagnostic diagnostic) = 0;
```
Comment (terpstra) [inline src/frontend/diagnostic.h:26]: I think you should define a helper method here:
```
void report(Location location, Severity severity, std::string message) { report(Diagnostic(location, severity, std::move(message)); }
```

File: src/frontend/parser.cpp [hunk 4]
```diff
@@ -1800,4 +1889,4 @@ Expr *parse_command(Lexer &lex) {
   auto out = parse_block(lex, false);
   expect(END, lex);
   return out;
-}
+}
```
Comment (terpstra) [inline src/frontend/parser.cpp:1892]: You removed the newline on this file.

File: src/frontend/parser.cpp [hunk 5]
```diff
@@ -26,18 +26,27 @@
...
 #include "frontend/expr.h"
 #include "frontend/symbol.h"
+#include "frontend/diagnostic.h"
 #include "runtime/value.h"
 #include "location.h"
 
 //#define TRACE(x) do { fprintf(stderr, "%s\n", x); } while (0)
 #define TRACE(x) do { } while (0)
 
+extern DiagnosticReporter *reporter;
```
Comment (terpstra) [inline src/frontend/parser.cpp:36]: Does a simple `extern DiagnosticReporter` work?

**Review by terpstra — APPROVED** [review]
Pre-approving contingent on fixes.

File: src/frontend/diagnostic.h (hunk 2 above)
Comment (elena-pudovkina) [reply src/frontend/diagnostic.h:29]: Fixed

File: src/frontend/parser.cpp (hunk 4 above)
Comment (elena-pudovkina) [reply src/frontend/parser.cpp:1892]: Sorry. Fixed it.

File: src/frontend/parser.cpp (hunk 5 above)
Comment (elena-pudovkina) [reply src/frontend/parser.cpp:36]: You can't create an object of an abstract class.

File: src/frontend/diagnostic.h (hunk 3 above)
Comment (elena-pudovkina) [reply src/frontend/diagnostic.h:26]: Done

File: src/cli/main.cpp (hunk 1 above)
Comment (elena-pudovkina) [reply src/cli/main.cpp:62]: Have to resort to using a pointer because DiagnosticReporter is an abstract class

**Pull request merged**